xarray/conventions.py

```diff
@@ -108,6 +108,10 @@ def ensure_dtype_not_object(var: Variable, name: T_Name = None) -> Variable:
     if var.dtype.kind == "O":
         dims, data, attrs, encoding = _var_as_tuple(var)
 
+        # leave vlen dtypes unchanged
```

Comment:
Actual fix by @tomwhite to preserve the dtype for vlen strings and not convert to float.